{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ostensibly examining VCF files with pyvcf\n",
    "### But really, accessing remote files and some bioinformatic dictionary usage\n",
    "\n",
    "Objectives:\n",
    "\n",
    "* access remote files with streams or by downloading\n",
    "* practise scanning through a file, in this case record by record of VCF files\n",
    "* use dictionaries both to access INFO fields and to generate summary data\n",
    "* Compare across dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import vcf\n",
    "import urllib.request\n",
    "import matplotlib.pylab as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The version of pyvcf we're using here doesn't handle opening remote URLs as nicely as pandas does.  However,\n",
    "it can take two types of input arguments; either something that looks like an open file, or a local filename.\n",
    "\n",
    "`urllib` lets us deal with remote files with URLs very nicely in a couple of ways we'll look at:\n",
    "scanning through the file, or just downloading it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NIST Genome in a Bottle NA12878 vcf/bed file repository (www.genomeinabottle.org)\n",
      "\n",
      "\n",
      "\n",
      "This is a repository of integrated snp and indel calls for NA12878, along with bed file that includes regions in which we believe our genotype calls are highly accurate (including homozygous reference calls if no snp or indel is called).  This is an updated version of our calls from Zook et al, Nat Biotech, 2014 (http://www.nature.com/nbt/journal/v32/n3/full/nbt.2835.html), which fixes a bug that caused more indels to be uncertain than necessary, so that these calls contain significantly more indels.  Otherwise, methods are similar to those described in the manuscript.\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "remote_readme=\"ftp://ftp-trace.ncbi.nih.gov/giab/ftp/release/NA12878_HG001/latest/README.NIST.v2.19.txt\"\n",
    "with urllib.request.urlopen(remote_readme) as remote_stream:\n",
    "    linecount = 0\n",
    "    for line in remote_stream:\n",
    "        #print(line)\n",
    "        print(line.decode('ascii'))\n",
    "        linecount = linecount + 1\n",
    "        if linecount > 3:\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can open such a stream with a vcf reader and get the header information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant Level information\n",
      "AC\n",
      "AF\n",
      "AN\n",
      "DP\n",
      "DPSum\n",
      "Entropy\n",
      "HRun\n",
      "HapNoVar\n",
      "LEN\n",
      "NoCG\n",
      "NoPLTot\n",
      "PL454WG\n",
      "PLCG\n",
      "PLHSWEx\n",
      "PLHSWG\n",
      "PLILL250\n",
      "PLILLCLIA\n",
      "PLILLWEx\n",
      "PLILLWG\n",
      "PLIllPCRFree\n",
      "PLIonEx\n",
      "PLPlatGen\n",
      "PLXIll\n",
      "PLXPSolWGLS\n",
      "PLminsum\n",
      "PLminsumOverDP\n",
      "RPA\n",
      "RU\n",
      "TYPE\n",
      "TrancheABQDmin2\n",
      "TrancheAlignmin2\n",
      "TrancheMapmin2\n",
      "TrancheSSEmin2\n",
      "YesPLtot\n",
      "allalts\n",
      "datasetcalls\n",
      "filter\n",
      "geno\n",
      "genoMapGood\n",
      "platformbias\n",
      "platformnames\n",
      "platforms\n",
      "varType\n",
      "Sample Level information\n",
      "DP\n",
      "GQ\n",
      "GT\n",
      "PL\n"
     ]
    }
   ],
   "source": [
    "remote_file = \"ftp://ftp-trace.ncbi.nih.gov/giab/ftp/release/NA12878_HG001/latest/NISTIntegratedCalls_14datasets_131103_allcall_UGHapMerge_HetHomVarPASS_VQSRv2.19_2mindatasets_5minYesNoRatio_all_nouncert_excludesimplerep_excludesegdups_excludedecoy_excludeRepSeqSTRs_noCNVs.vcf.gz\"\n",
    "with urllib.request.urlopen(remote_file) as remote_stream:\n",
    "    v = vcf.Reader(remote_stream, compressed=True)\n",
    "\n",
    "    print('Variant Level information')\n",
    "    infos = v.infos\n",
    "    for info in infos:\n",
    "        print(info)\n",
    "\n",
    "    print('Sample Level information')\n",
    "    fmts = v.formats\n",
    "    for fmt in fmts:\n",
    "        print(fmt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But because we'll be scanning over this file repeatedly, let's get a subset of it locally:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NIST_chr20.vcf.gz', <http.client.HTTPMessage at 0x109891780>)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr20_filename=\"https://github.com/ljdursi/mon_py_wrkshp/blob/master/NISTIntegratedCalls_chr20.vcf.gz?raw=true\"\n",
    "localfilename=\"NIST_chr20.vcf.gz\"\n",
    "urllib.request.urlretrieve(chr20_filename, localfilename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see what we can do with one of these files.  When we open a vcf file, we have to call a function `next` to get a record:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=20, POS=61138, REF=C, ALT=[CT])\n",
      "{'RPA': [12, 13], 'platforms': 1, 'PLminsum': 4173, 'HapNoVar': 0, 'PLILLWG': ['133', '0', '108', '133', '133', '133'], 'PLNCIIonWG': ['0', '127', '127', '18', '127', '127'], 'PLILLCLIA': ['512', '0', '657', '657', '657', '657'], 'TrancheSSEmin2': 90.0, 'platformbias': ['none'], 'TrancheABQDmin2': 90.0, 'datasetcalls': 10, 'PLPlatGen': ['1922', '0', '2315', '2315', '2315', '2315'], 'PLIonEx': ['filteredSSE99.9Align99.9', '0', '46', '46', '6', '46', '46'], 'TrancheAlignmin2': 0.0, 'platformnames': ['ill'], 'varType': 'INDEL', 'PLHSWG': ['684', '0', '649', '684', '684', '684'], 'PLIllPCRFree': ['466', '0', '430', '466', '466', '466'], 'HRun': 0, 'YesPLtot': 7, 'geno': 2, 'RU': 'T', 'PLminsumOverDP': 7.31, 'PLXIll': ['281', '0', '284', '284', '284', '284'], 'genoMapGood': 7, 'allalts': ['CT', 'T'], 'DPSum': 571, 'NoPLTot': 1, 'PLCG': ['6', '0', '100', '100', '100', '100'], 'TrancheMapmin2': 0.0, 'PLILL250': ['296', '0', '430', '430', '430', '430']}\n",
      "['ill']\n",
      "[Call(sample=NA12878, CallData(GT=0/1, DP=571, GQ=4173, PL=[4300, 127, 5100]))]\n",
      "0/1\n"
     ]
    }
   ],
   "source": [
    "reader = vcf.Reader(filename=localfilename)\n",
    "record = next(reader)\n",
    "print(record)\n",
    "print(record.INFO)\n",
    "print(record.INFO['platformnames'])\n",
    "print(record.samples)\n",
    "print(record.samples[0]['GT'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The usual way to process a file, VCF or otherwise, is to loop over lines in the file and process them as you like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=20, POS=61138, REF=C, ALT=[CT]) 0/1\n",
      "Record(CHROM=20, POS=67500, REF=T, ALT=[TTGGTATCTAG]) 0/1\n",
      "Record(CHROM=20, POS=68749, REF=T, ALT=[C]) 1/1\n",
      "Record(CHROM=20, POS=70484, REF=CTCTT, ALT=[C]) 0/1\n",
      "Record(CHROM=20, POS=70980, REF=G, ALT=[A]) 0/1\n"
     ]
    }
   ],
   "source": [
    "reader = vcf.Reader(filename=localfilename)\n",
    "recordnum = 0\n",
    "for record in reader:\n",
    "    print(record, record.samples[0]['GT'])\n",
    "    recordnum += 1\n",
    "    if recordnum == 5:\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hands-on: write a function, `get_depths`, which is given a filename and generates a list of all `DPSum` info fields - this is the total depth across all platforms at which this base was seen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_depths(input_file):\n",
    "    depth_counts = []\n",
    "    depth_item = 'DPSum'\n",
    "    reader = vcf.Reader(filename=input_file)\n",
    "    for record in reader:\n",
    "        if depth_item in record.INFO:\n",
    "            depth = int(record.INFO[depth_item])\n",
    "            depth_counts.append(depth)\n",
    "    return depth_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "depths = get_depths(localfilename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([  5.00000000e+00,   7.00000000e+00,   1.30000000e+01,\n",
       "          1.70000000e+01,   2.90000000e+01,   3.30000000e+01,\n",
       "          6.00000000e+01,   9.60000000e+01,   1.05000000e+02,\n",
       "          1.76000000e+02,   2.13000000e+02,   3.31000000e+02,\n",
       "          4.50000000e+02,   5.55000000e+02,   8.07000000e+02,\n",
       "          1.07400000e+03,   1.47600000e+03,   2.01000000e+03,\n",
       "          2.66600000e+03,   3.63600000e+03,   5.19100000e+03,\n",
       "          6.66000000e+03,   9.00000000e+03,   9.50300000e+03,\n",
       "          9.59300000e+03,   7.85300000e+03,   5.22900000e+03,\n",
       "          3.10800000e+03,   1.52900000e+03,   7.44000000e+02,\n",
       "          3.66000000e+02,   1.54000000e+02,   1.08000000e+02,\n",
       "          9.00000000e+01,   6.50000000e+01,   6.00000000e+01,\n",
       "          5.20000000e+01,   4.60000000e+01,   3.20000000e+01,\n",
       "          1.80000000e+01,   1.70000000e+01,   1.50000000e+01,\n",
       "          3.00000000e+00,   5.00000000e+00,   3.00000000e+00,\n",
       "          0.00000000e+00,   3.00000000e+00,   2.00000000e+00,\n",
       "          3.00000000e+00,   2.00000000e+00]),\n",
       " array([   92.  ,   117.62,   143.24,   168.86,   194.48,   220.1 ,\n",
       "          245.72,   271.34,   296.96,   322.58,   348.2 ,   373.82,\n",
       "          399.44,   425.06,   450.68,   476.3 ,   501.92,   527.54,\n",
       "          553.16,   578.78,   604.4 ,   630.02,   655.64,   681.26,\n",
       "          706.88,   732.5 ,   758.12,   783.74,   809.36,   834.98,\n",
       "          860.6 ,   886.22,   911.84,   937.46,   963.08,   988.7 ,\n",
       "         1014.32,  1039.94,  1065.56,  1091.18,  1116.8 ,  1142.42,\n",
       "         1168.04,  1193.66,  1219.28,  1244.9 ,  1270.52,  1296.14,\n",
       "         1321.76,  1347.38,  1373.  ]),\n",
       " <a list of 50 Patch objects>)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAEACAYAAABPiSrXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE9VJREFUeJzt3W+snOV55/HvDzuEhBDHqWQf1SaYlpiYardZV+tkN1vl\nqIkgpJLhxS4lrRoIqC8adhOlVRs7fYHzKgEpKqm2IEVNwaS0lJBd4d1ScC16WlVaFreENcEueEUx\ntrM+qE1rKV0JQXL1xTyE4fgcc/vMjGfG5/uRRjxzzfPMucbMnN/c9/PnpKqQJKnFeeNuQJI0PQwN\nSVIzQ0OS1MzQkCQ1MzQkSc0MDUlSszcNjSRfTzKf5EBfbW2SvUmeTfJokjV9j+1McjjJoSRX9tW3\nJjmQ5Lkkd/TVz09yf7fN/0rynmG+QEnS8LSMNO4GrlpQ2wHsq6rLgceAnQBJrgCuA7YAVwN3Jkm3\nzV3AzVW1Gdic5LXnvBn4XlW9F7gDuH2A1yNJGqE3DY2q+ivgHxeUrwF2d8u7gWu75e3A/VX1alW9\nABwGtiWZAS6qqv3devf2bdP/XA8CH1nG65AknQXL3aexrqrmAarqBLCuq28Ajvatd7yrbQCO9dWP\ndbU3bFNVPwD+Kcm7l9mXJGmEhrUjfJjXIsmbryJJGofVy9xuPsn6qprvpp5e6urHgYv71tvY1Zaq\n92/z3SSrgHdW1fcW+6FJvFCWJC1DVQ3lC3nrSCO8cQSwB7ixW74BeKivfn13RNSlwGXAE90U1skk\n27od459csM0N3fJ/ordjfUlVNbW3W2+9dew92P/4+1hpvdv/+G/D9KYjjSR/CMwCP5bkReBW4MvA\nN5PcBByhd8QUVXUwyQPAQeAV4NP1ese3APcAFwAPV9UjXf3rwDeSHAb+Abh+OC9NkjRsbxoaVfWL\nSzz00SXW/xLwpUXqfwP8q0XqL9OFjiRpsnlG+Fk0Ozs77hYGYv/jM829g/2fSzLs+a5RSlLT1K8k\nTYIk1FneES5JkqEhSWpnaEiSmhka0hDNzGwiySm3mZlN425NGgp3hEtD1Dt3dbH3aIZ+kpXUyh3h\n0pgtNaKQznWONKRlON2IwpGGJo0jDUnSWBgakqRmhoYkqZmhIUlqZmhIkpoZGpKkZoaGJKmZoSFJ\namZoSJKaGRrSaXi5EOmNvIyIdBrLuVyIlxHRpPEyIpKksTA0JEnNDA1JUjNDQ5LUzNCQJDUzNCRJ\nzQwNSVIzQ0OS1MzQkCQ1MzQkSc0MDUlSM0NDktTM0JAkNTM0JEnNDA1JUjNDQ5LUbKDQSPK5JN9J\nciDJfUnOT7I2yd4kzyZ5NMmavvV3Jjmc5FCSK/vqW7vneC7JHYP0JEkanWWHRpIfB/4LsLWq/jWw\nGvgEsAPYV1WXA48BO7v1rwCuA7YAVwN35vW/m3kXcHNVbQY2J7lquX1JkkZn0OmpVcCFSVYDbwOO\nA9cAu7vHdwPXdsvbgfur6tWqegE4DGxLMgNcVFX7u/Xu7dtGkjRBlh0aVfVd4CvAi/TC4mRV7QPW\nV9V8t84JYF23yQbgaN9THO9qG4BjffVjXU2SNGFWL3fDJO+iN6q4BDgJfDPJLwG1YNWF9weya9eu\nHy3Pzs4yOzs7zKeXpKk3NzfH3NzcSJ47Vcv7nZ7kPwJXVdWvdPd/Gfgg8HPAbFXNd1NPf15VW5Ls\nAKqqbuvWfwS4FTjy2jpd/Xrgw1X1q4v8zFpuv9Jy9Ha7LfaeO/O6712NSxKqKm++5psbZJ/Gi8AH\nk1zQ7dD+CHAQ2APc2K1zA/BQt7wHuL47wupS4DLgiW4K62SSbd3zfLJvG0nSBFn29FRVPZHkQeDb\nwCvdf78GXAQ8kOQmeqOI67r1DyZ5gF6wvAJ8um/YcAtwD3AB8HBVPbLcviRJo7Ps6alxcHpKZ9vw\npqcuAF4+pbp+/SWcOPHCAB1Kb26Y01OGhnQaw9yn4b4Ojcuk7NOQJK0whoYkqZmhIUlqZmhIkpoZ\nGpKkZoaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhSWpmaEiSmhkakqRmhoYkqZmhIUlqZmhIkpoZ\nGpKkZoaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhSWpmaEiSmhkaEjAzs4kkp9wkvVGqatw9NEtS\n09SvpkcvIBZ7b42+7ntao5aEqhrKtyBHGpKkZoaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhSWpm\naEiSmg0UGknWJPlmkkNJnknygSRrk+xN8mySR5Os6Vt/Z5LD3fpX9tW3JjmQ5LkkdwzSkyRpdAYd\naXwVeLiqtgA/DfwtsAPYV1WXA48BOwGSXAFcB2wBrgbuzOvXabgLuLmqNgObk1w1YF+SpBFYdmgk\neSfws1V1N0BVvVpVJ4FrgN3daruBa7vl7cD93XovAIeBbUlmgIuqan+33r1920iSJsggI41Lgb9P\ncneSJ5N8LcnbgfVVNQ9QVSeAdd36G4Cjfdsf72obgGN99WNdTZI0YQYJjdXAVuB3q2or8M/0pqYW\nXn3Nq7FJ0jli9QDbHgOOVtVfd/e/RS805pOsr6r5burppe7x48DFfdtv7GpL1Re1a9euHy3Pzs4y\nOzs7wEuQpHPP3Nwcc3NzI3nugS6NnuQvgF+pqueS3Aq8vXvoe1V1W5LPA2urake3I/w+4AP0pp/+\nDHhvVVWSx4HPAPuBPwF+p6oeWeTneWl0jYSXRte5bJiXRh9kpAG9X/T3JXkL8DzwKWAV8ECSm4Aj\n9I6YoqoOJnkAOAi8Any6LwFuAe4BLqB3NNYpgSFJGj//CJOEIw2d2/wjTJKksTA0JEnNDA1JUjND\nQ5LUzNCQJDUzNCRJzQwNSVIzQ0OS1MzQkCQ1MzQkSc0MDUlSM0NDktTM0JAkNTM0pLF6K0kWvc3M\nbBp3c9IpDA2tKDMzmxb9BT0+L9O7ZPqpt/n5I2PsS1qcf09DK8o4/27GmdV7j/l+1zD49zQkSWNh\naEiSmhkakqRmhoYkqZmhIUlqZmhIkpoZGpKkZoaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhSWpm\naEiSmhkakqRmhoYkqZmhIUlqZmhIkpoZGpKkZoaGJKnZwKGR5LwkTybZ091fm2RvkmeTPJpkTd+6\nO5McTnIoyZV99a1JDiR5Lskdg/YkSRqNYYw0Pgsc7Lu/A9hXVZcDjwE7AZJcAVwHbAGuBu5Mkm6b\nu4Cbq2ozsDnJVUPoS5I0ZAOFRpKNwMeB3+srXwPs7pZ3A9d2y9uB+6vq1ap6ATgMbEsyA1xUVfu7\n9e7t20aSNEEGHWn8NvAbQPXV1lfVPEBVnQDWdfUNwNG+9Y53tQ3Asb76sa4mSZowyw6NJD8PzFfV\nU0BOs2qd5jFJ0hRZPcC2HwK2J/k48DbgoiTfAE4kWV9V893U00vd+seBi/u239jVlqovateuXT9a\nnp2dZXZ2doCXIEnnnrm5Oebm5kby3KkafCCQ5MPAr1fV9iS3A/9QVbcl+Tywtqp2dDvC7wM+QG/6\n6c+A91ZVJXkc+AywH/gT4Heq6pFFfk4No1+tXL1jLxZ7D01avfeY73cNQxKq6nQzQs0GGWks5cvA\nA0luAo7QO2KKqjqY5AF6R1q9Any6LwFuAe4BLgAeXiwwJEnjN5SRxtniSEODcqShlWiYIw3PCJck\nNTM0JEnNDA1JUjNDQ5LUzNDQOWdmZhNJFr1JGoxHT+mcs/QRUjB5R0l59JRGz6OnJEljYWhIkpoZ\nGpKkZoaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhSWpmaEgT662LXgplZmbTuBvTCuZlRHTOOZcu\nI7LUNn4OdCa8jIgkaSwMDUlSM0NDktTM0JAkNTM0JEnNDA1JUjNDQ5LUzNCQJDUzNDS1ZmY2LXrG\ntKTR8YxwTa2lz/z2jHCpn2eES5LGwtCQJDUzNCRJzQwNSVIzQ0OS1MzQkCQ1MzQkSc0MDUlSM0ND\nktRs2aGRZGOSx5I8k+TpJJ/p6muT7E3ybJJHk6zp22ZnksNJDiW5sq++NcmBJM8luWOwlyRJGpVB\nRhqvAr9WVT8F/DvgliTvA3YA+6rqcuAxYCdAkiuA64AtwNXAnXn9QkF3ATdX1WZgc5KrBuhLkjQi\nyw6NqjpRVU91y98HDgEbgWuA3d1qu4Fru+XtwP1V9WpVvQAcBrYlmQEuqqr93Xr39m0jSZogQ9mn\nkWQT8H7gcWB9Vc1DL1iAdd1qG4CjfZsd72obgGN99WNdTZI0YVYP+gRJ3gE8CHy2qr6fZOHlN4d6\nOc5du3b9aHl2dpbZ2dlhPr0m0MzMJubnj4y7DWlqzM3NMTc3N5LnHujS6ElWA/8T+NOq+mpXOwTM\nVtV8N/X051W1JckOoKrqtm69R4BbgSOvrdPVrwc+XFW/usjP89LoK9CZXwLdS6NL/Sbp0ui/Dxx8\nLTA6e4Abu+UbgIf66tcnOT/JpcBlwBPdFNbJJNu6HeOf7NtGkjRBlj3SSPIh4C+Bp+l9HSrgC8AT\nwAPAxfRGEddV1T912+wEbgZeoTedtber/wxwD3AB8HBVfXaJn+lIYwVypHFq3c+BzsQwRxr+5T5N\nPEPj1LqfA52JSZqekiStIIaGJKmZoSFJamZoSJKaGRqSpGaGhiSpmaEhTZ23kuSU28zMpnE3phXA\n0NDEmJnZtOgvQy30Mq+fT/v6zetz6Wzw5D5NjOGdxHfun9znSX86E57cJ0kaC0NDktTM0JAkNTM0\nJEnNDA1JUjNDQ5LUzNCQJDUzNHTWeRKfNL08uU9n3ehP4vPkPqmfJ/dJksbC0JAkNTM0JEnNDA1J\nUjNDQyPjUVLSucfQ0Mj0/r7DqX/3QaPiH2fS6HnIrUZmfIfWrtxDbj0UV4vxkFtJ0lgYGpKkZoaG\nBuYOb2nlcJ+GBjZ5+y7cp7Gw7udmZXOfhiRpLAwNNXMaSpLTU2o2PdNQTk8trPu5WdmcntLILDWa\ncEQxzTzpT8PjSENvsPRoAibv27gjjUHrfp5WBkcakqSxmJjQSPKxJH+b5Lkknx93P+c6d2pLWo6J\nCI0k5wH/FbgK+CngE0neN96uhm9ubu6s/8ylwmFlXkxwbtwNDGBuBM959vZ1jOO9P0zT3v8wTURo\nANuAw1V1pKpeAe4HrhlzT0M3yjee4dBibtwNDGBuBM/5Mou9N+bnTww9TKb9l+609z9MkxIaG4Cj\nffePdbUVa6kQWLXqQsNBI3ZmYbLUe7L/sS9+8YseuXWOmJTQGJv9+/cv+Ya//favnNFzne5w1VWr\nLjzlg7OcEPjhD///onVp9BYPk6Xfk/2P3cogATTquiHWbiIOuU3yQWBXVX2su78DqKq6bcF6429W\nkqbQsA65nZTQWAU8C3wE+H/AE8AnqurQWBuTJL3B6nE3AFBVP0jyn4G99KbMvm5gSNLkmYiRhiRp\nOkzNjvBJP/kvycYkjyV5JsnTST7T1dcm2Zvk2SSPJlnTt83OJIeTHEpy5fi6f12S85I8mWRPd39q\n+k+yJsk3u36eSfKBaek/yeeSfCfJgST3JTl/0ntP8vUk80kO9NXOuOckW7vX/VySO8bY++1db08l\n+VaSd05i70v13/fYryf5YZJ3j6T/qpr4G71w+7/AJcBbgKeA9427rwU9zgDv75bfQW8fzfuA24Df\n7OqfB77cLV8BfJveFOGm7vVlAl7H54A/APZ096emf+Ae4FPd8mpgzTT0D/w48Dxwfnf/j4EbJr13\n4D8A7wcO9NXOuGfgfwP/tlt+GLhqTL1/FDivW/4y8KVJ7H2p/rv6RuAR4O+Ad3e1LcPsf1pGGhN/\n8l9Vnaiqp7rl7wOH6P0PvAbY3a22G7i2W94O3F9Vr1bVC8Bheq9zbJJsBD4O/F5feSr6774V/mxV\n3Q3Q9XWSKekfWAVcmGQ18DbgOBPee1X9FfCPC8pn1HOSGeCiqtrfrXdv3zYjs1jvVbWvqn7Y3X2c\n3ud34nrvel3s3x7gt4HfWFC7hiH2Py2hMVUn/yXZRO9bwOPA+qqah16wAOu61Ra+puOM/zW99obr\n39E1Lf1fCvx9kru76bWvJXk7U9B/VX0X+ArwYtfHyaraxxT0voh1Z9jzBnqf59dMymf7JnrfvGFK\nek+yHThaVU8veGio/U9LaEyNJO8AHgQ+2404Fh5pMJFHHiT5eWC+Gy2d7njuieyf3tB7K/C7VbUV\n+GdgB1Pw75/kXfS+DV5Cb6rqwiS/xBT03mDqek7yW8ArVfVH4+6lVZK3AV+gdxblSE1LaBwH3tN3\nf2NXmyjd1MKDwDeq6qGuPJ9kfff4DPBSVz8OXNy3+bhf04eA7UmeB/4I+Lkk3wBOTEn/x+h9y/rr\n7v636IXINPz7fxR4vqq+V1U/AP478O+Zjt4XOtOeJ+q1JLmR3hTtL/aVp6H3n6S3v+L/JPm7rpcn\nk6xj6d+fy+p/WkJjP3BZkkuSnA9cD+wZc0+L+X3gYFV9ta+2B7ixW74BeKivfn13lMylwGX0Tmoc\ni6r6QlW9p6p+gt6/72NV9cvA/2A6+p8HjibZ3JU+AjzDdPz7vwh8MMkFSUKv94NMR+/hjSPTM+q5\nm8I6mWRb99o/2bfNqL2h9yQfozc9u72qXu5bbxJ7h77+q+o7VTVTVT9RVZfS+xL1b6rqpa7/Xxha\n/2djT/+Qjhb4GL0jkg4DO8bdzyL9fQj4Ab0ju74NPNn1/G5gX9f7XuBdfdvspHckwyHgynG/hr6+\nPszrR09NTf/AT9P7gvEU8N/oHT01Ff3Tm1Y4BBygtwP5LZPeO/CHwHfpXZTqReBTwNoz7Rn4GeDp\n7rP91TH2fhg40n12nwTunMTel+p/wePP0x09Nez+PblPktRsWqanJEkTwNCQJDUzNCRJzQwNSVIz\nQ0OS1MzQkCQ1MzQkSc0MDUlSs38B0oiTD2OgjDQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x112646be0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(depths, bins=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO:\n",
    "\n",
    "* Demonstrate: get counts as dictionary\n",
    "    * defaultdict\n",
    "* Hands on: generate defaultdict of counts by varianttype, genotype\n",
    "* Generate list of dictionaries of calls by platform type; using those, investigate overlap and no-overlap of calls by platform.  Eg, how many calls were made by illumina that weren't made by 10x?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
